### PR TITLE
Chapter1/homework/4

### PR DESCRIPTION
**Pull request by BootNinja**
All Tests pass.

### CODE REVIEW

File: Src/BootCamp.Chapter/Lesson3.cs [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+            }
+            
+            if (height < 0 || weight<=0)
+            {
+                Console.WriteLine("Failed calculating BMI. Reason:");
+                
+                if (weight <= 0)
+                    Console.WriteLine("Weight cannot be equal or less than zero, but was " + weight + ".");
+               
+                if (height <= 0 && weight>=0)
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Lesson3.cs]: Codacy Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886326)

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 5
           

Complexity increasing per file
==============================
- Src/BootCamp.Chapter/Lesson3.cs  8
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886326&bid=16186543)

File: Src/BootCamp.Chapter/Lesson3.cs [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+                return parsedfloat;
+            }
+
+            //default case
+            BadInput(value);
+            return -1;
+
+        }
+        // Input function
+        public static void GatherData(out string firstName, out string lastName, out int age, out float weight, out float height)
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Lesson3.cs]: Codacy Issue found: [Consider refactoring this method in order to remove the need for this 'out' modifier.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886326)

File: Src/BootCamp.Chapter/Lesson3.cs [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+            BadInput(value);
+            return -1;
+            
+        }
+        public static float GetFloatInput(string prompt)
+        {
+            Console.WriteLine(prompt);
+            string value = Console.ReadLine();
+
+            if (Empty(value))
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Lesson3.cs]: Codacy Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886326)

File: Src/BootCamp.Chapter/Lesson3.cs [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+           
+
+
+
+            return weight / (height * height);
+        }
+
+        public static void BadInput(string value)
+        {
+            Console.Write("\"" + value.ToString() + "\" is not a valid number.");
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Lesson3.cs]: Codacy Issue found: [There's no need to call 'ToString()' on a string.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886326)

File: Src/BootCamp.Chapter/Lesson3.cs [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+                Console.WriteLine("Height cannot be equal or less than zero, but was " + height + ".");
+
+                return -1F;
+            }
+            
+            if (height < 0 || weight<=0)
+            {
+                Console.WriteLine("Failed calculating BMI. Reason:");
+                
+                if (weight <= 0)
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Lesson3.cs]: Codacy Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886326)

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- Src/BootCamp.Chapter/Lesson3.cs  8
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886326&bid=16186543)

File: Src/BootCamp.Chapter/Lesson3.cs [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+            //default case
+            return value;
+        }
+        public static int GetIntInput(string prompt)
+        {
+            Console.WriteLine(prompt);
+            string value = Console.ReadLine();
+
+            if (Empty(value))
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Lesson3.cs]: Codacy Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886326)

File: Src/BootCamp.Chapter/Lesson3.cs [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+                return parsedfloat;
+            }
+
+            //default case
+            BadInput(value);
+            return -1;
+
+        }
+        // Input function
+        public static void GatherData(out string firstName, out string lastName, out int age, out float weight, out float height)
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Lesson3.cs]: Codacy Issue found: [Consider refactoring this method in order to remove the need for this 'out' modifier.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886326)

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- Src/BootCamp.Chapter/Lesson3.cs  8
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886326&bid=16186543)

File: Src/BootCamp.Chapter/Lesson3.cs [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Lesson3.cs]: Codacy Issue found: [Consider refactoring this method in order to remove the need for this 'out' modifier.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886326)

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- Src/BootCamp.Chapter/Lesson3.cs  8
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886326&bid=16186543)

File: Src/BootCamp.Chapter/Lesson3.cs [hunk 9]
```diff
@@ -0,0 +1,138 @@
+﻿using System;
+
+
+namespace BootCamp.Chapter
+{    public class Lesson3
+    {   public static void Demo()
```
Comment (ellmaster) [inline Src/BootCamp.Chapter/Lesson3.cs]: I don't think you wanted it to be like that? Either way, please change the format to fit to the normal styling.

File: Src/BootCamp.Chapter/Lesson3.cs [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+
+            return weight / (height * height);
+        }
+        //outputs error message if given invalid input
+        public static void BadInput(string value)
+        {
+            Console.Write("\"" + value + "\" is not a valid number.");
+        }
+        //returns true on an empty input
+        public static bool Empty(string value)
```
Comment (ellmaster) [inline Src/BootCamp.Chapter/Lesson3.cs]: Boolean functions (and variables) should express questions that they answer on.
So for example here you would want it to be at least IsEmpty(string value).

File: Src/BootCamp.Chapter/Lesson3.cs [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+        {
+            if (weight > 0 && height <= 0)
+            {
+                Console.WriteLine("Failed calculating BMI. Reason:");
+                Console.WriteLine("Height cannot be equal or less than zero, but was " + height + ".");
+
+                return -1F;
+            }
+            
+            if (height < 0 || weight<=0)
```
Comment (ellmaster) [inline Src/BootCamp.Chapter/Lesson3.cs]: Pay attention to the correct spacing throughout your document! :)

File: Src/BootCamp.Chapter/Lesson3.cs [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+                Console.ReadKey();
+            }
+        }
+        // Output function
+        public static void DisplayOutput(string first, string last, int age, float weight, float height)
+        {
+            Console.Clear();
+            Console.WriteLine(first + " " + last + " is " + age + " years old; his weight is " + weight + " kg and his height is " + (height * 100) + " cm.");
+            Console.WriteLine(first + "'s Body Mass Index is " + CalcBmi(weight, height) + ".");
+            Console.WriteLine("\n\n\n\n\nPress Enter");
```
Comment (ellmaster) [inline Src/BootCamp.Chapter/Lesson3.cs]: If you make use of NewLine characters more often you could think about creating a function like WriteNewLines(int count).

Comment (BootNinja) [reply Src/BootCamp.Chapter/Lesson3.cs]: agreed, but i felt that was overkill for this instance.

File: Src/BootCamp.Chapter/Lesson3.cs (hunk 11 above)
Comment (BootNinja) [reply Src/BootCamp.Chapter/Lesson3.cs]: corrected

File: Src/BootCamp.Chapter/Lesson3.cs (hunk 10 above)
Comment (BootNinja) [reply Src/BootCamp.Chapter/Lesson3.cs]: I would quibble on this one, but ok, done.

File: Src/BootCamp.Chapter/Lesson3.cs (hunk 9 above)
Comment (BootNinja) [reply Src/BootCamp.Chapter/Lesson3.cs]: not sure how that happened.  fixed.

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- Src/BootCamp.Chapter/Lesson3.cs  8
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886326&bid=16186543)

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- Src/BootCamp.Chapter/Lesson3.cs  8
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886326&bid=16186543)

File: Src/BootCamp.Chapter/Lesson3.cs (hunk 10 above)
Comment (ellmaster) [reply Src/BootCamp.Chapter/Lesson3.cs]: Feel free to do so!
Sometimes you need to have a conversation about such a thing to understand it right.

I will try to explain it that way:
Assume your project grows and grows and grows and you maybe have not a function like Empty but something less obvious. Now you (or maybe a 3rd party user) want(s) to look up a certain functionality.
By using HasX, IsX, etc. you can easily find such functions with a proper IDE like VS is. Also if you are used to that, you probably will not find Empty anymore and end up writing it again, maybe at some other place in your code.

I hope this makes my intentions behind that much more clear. If not - just ask again! :)

**Pull request closed without merging**